include/myst/thread.h

```diff
@@ -405,13 +438,66 @@ MYST_INLINE int myst_thread_queue_remove_thread(
     return found ? pos : -1;
 }
 
+/** Find and remove up to `n` threads from the queue such that
+ * <waiting bitset> & <passed in bitset> != 0.
+ * Removed threads are placed in the `matches` queue.
+ */
+MYST_INLINE size_t myst_thread_queue_search_remove_bitset(
+    myst_thread_queue_t* queue,
+    myst_thread_queue_t* matches,
+    size_t n,
+    uint32_t bitset)
+{
+    if (!queue || !matches)
+        return -1;
+
+    myst_thread_t* t = queue->_front;
+    myst_thread_t* prev = NULL;
+
+    size_t num_found = 0;
+
+    while (t)
+    {
+        if (num_found >= n)
+            break;
+
+        myst_thread_t* next = t->_qnext;
+        if (t->_bitset & bitset)
+        {
+            num_found++;
+            if (prev != NULL)
+            {
+                prev->_qnext = next;
+                if (next == NULL)
+                    queue->_back = prev;
+            }
+            else
+            {
+                queue->_front = queue->_front->_qnext;
```

Comment:
Isn't `queue->_front->_qnext` the same as `next`? If so, is 477 & 478 the same as 471 & 472?

Reply:
This handles removing the only node
```
if (queue->_front == NULL)
                    queue->_back = NULL;
```

This handles removing the last node
```
if (next == NULL)
                    queue->_back = prev;

```

Reply:
I simplified the cases.